base_layer/core/src/base_node/service/service.rs

```diff
@@ -280,222 +309,223 @@ where B: BlockchainBackend + 'static
         }
         Ok(())
     }
+}
 
-    async fn handle_incoming_request(
-        &mut self,
-        domain_request_msg: DomainMessage<proto::BaseNodeServiceRequest>,
-    ) -> Result<(), BaseNodeServiceError>
-    {
-        let (origin_public_key, inner_msg) = domain_request_msg.into_origin_and_inner();
-
-        // Convert proto::BaseNodeServiceRequest to a BaseNodeServiceRequest
-        let request = inner_msg
-            .request
-            .ok_or_else(|| BaseNodeServiceError::InvalidRequest("Received invalid base node request".to_string()))?;
-
-        let response = self
-            .inbound_nch
-            .handle_request(&request.try_into().map_err(BaseNodeServiceError::InvalidRequest)?)
-            .await?;
-
-        let message = proto::BaseNodeServiceResponse {
-            request_key: inner_msg.request_key,
-            response: Some(response.into()),
-        };
-
-        self.outbound_message_service
-            .send_direct(
-                origin_public_key,
-                OutboundEncryption::EncryptForPeer,
-                OutboundDomainMessage::new(TariMessageType::BaseNodeResponse, message),
-            )
-            .await?;
-
-        Ok(())
-    }
-
-    async fn handle_incoming_response(
-        &mut self,
-        incoming_response: proto::BaseNodeServiceResponse,
-    ) -> Result<(), BaseNodeServiceError>
-    {
-        let proto::BaseNodeServiceResponse { request_key, response } = incoming_response;
-
-        let mut finalize_request = false;
-        match self.waiting_requests.get_mut(&request_key) {
-            Some(waiting_request) => {
-                let response = response.and_then(|r| r.try_into().ok()).ok_or_else(|| {
-                    BaseNodeServiceError::InvalidResponse("Received an invalid base node response".to_string())
-                })?;
-                waiting_request.received_responses.push(response);
```

Comment:
Possible contention point

Reply:
This was changed, a single write lock is acquired and used for both the push and remove operations.